python-checks/src/main/java/org/sonar/python/checks/DictionaryDuplicateKeyCheck.java

```diff
@@ -0,0 +1,132 @@
+      for (int i = 0; i < dictionaryLiteral.elements().size(); i++) {
+        if (!dictionaryLiteral.elements().get(i).is(Tree.Kind.KEY_VALUE_PAIR) || issueIndexes.contains(i)) {
+          continue;
+        }
+        Expression key = ((KeyValuePair) dictionaryLiteral.elements().get(i)).key();
+        List<Tree> duplicateKeys = findIdenticalKeys(i, dictionaryLiteral.elements(), issueIndexes);
+        if (!duplicateKeys.isEmpty()) {
+          PreciseIssue issue = ctx.addIssue(key, "Change or remove duplicates of this key.");
+          duplicateKeys.forEach(d -> issue.secondary(d, "Duplicate key"));
+        }
+      }
+    });
+  }
+
+  private List<Tree> findIdenticalKeys(int startIndex, List<DictionaryLiteralElement> elements, Set<Integer> issueIndexes) {
+    Expression key = ((KeyValuePair) elements.get(startIndex)).key();
+    List<Tree> duplicates = new ArrayList<>();
+    for (int i = startIndex + 1; i < elements.size() ; i++) {
+      if (!elements.get(i).is(Tree.Kind.KEY_VALUE_PAIR)) {
+        continue;
+      }
+      Expression comparedKey = ((KeyValuePair) elements.get(i)).key();
+      if (isSameKey(key, comparedKey)) {
+        issueIndexes.add(i);
+        duplicates.add(comparedKey);
+      }
+    }
+    return duplicates;
+  }
+
+  private boolean isSameKey(Expression key, Expression comparedKey) {
+    if (key.is(Tree.Kind.STRING_LITERAL) && comparedKey.is(Tree.Kind.STRING_LITERAL)) {
+      return areEquivalentStringLiterals((StringLiteral) key, (StringLiteral) comparedKey);
+    }
+    if (isANumber(key) && isANumber(comparedKey)) {
+      return areEquivalentNumbers(key, comparedKey);
+    }
+    return !key.is(Tree.Kind.CALL_EXPR) && CheckUtils.areEquivalent(key, comparedKey);
+  }
+
+  private boolean areEquivalentNumbers(Tree key, Tree comparedKey) {
+    return toBigDecimal(key).compareTo(toBigDecimal(comparedKey)) == 0;
+  }
+
+  private BigDecimal toBigDecimal(Tree numberTree) {
+    if (numberTree.is(Tree.Kind.NUMERIC_LITERAL)) {
+      return parseAsBigDecimal(((NumericLiteral) numberTree).valueAsString());
+    }
+    return ((Name) numberTree).name().equals("True") ? BigDecimal.ONE : BigDecimal.ZERO;
+  }
+
+  private static boolean areEquivalentStringLiterals(StringLiteral key, StringLiteral comparedKey) {
+    if (key.stringElements().stream().anyMatch(StringElement::isInterpolated) || comparedKey.stringElements().stream().anyMatch(StringElement::isInterpolated)) {
+      return false;
+    }
+    if (key.trimmedQuotesValue().equals(comparedKey.trimmedQuotesValue())) {
+      String keyWithPrefixes = key.stringElements().stream()
+        .map(s -> s.prefix().toLowerCase(Locale.ENGLISH) + s.trimmedQuotesValue()).collect(Collectors.joining());
+      String comparedKeyWithPrefixes = comparedKey.stringElements().stream()
+        .map(s -> s.prefix().toLowerCase(Locale.ENGLISH) + s.trimmedQuotesValue()).collect(Collectors.joining());
+      return keyWithPrefixes.equals(comparedKeyWithPrefixes);
```

Comment:
`{ "a" r"b": 'y', r"ab": 'n' }` gives a false negative, because `"arb" != "rab"`, but that's, admittedly, a rather obscure corner case. In order to improve that, a more principled approach for string literal interpretation would be needed (probably out of scope).

Reply:
Indeed!
In order to fix this we would need to resolve the actual result of the string literal, which is out of scope for this rule. This heuristic was chosen to cover the basic cases of implicit concatenation and prefix usages without risking to raise FPs.